pkg/cmd/serve.go

```diff
@@ -50,116 +53,142 @@ func RegisterServeFlags(cmd *cobra.Command, config *server.Config) error {
+	if err := namespaceCacheFlags.MarkHidden("ns-cache-expiration"); err != nil {
 		return fmt.Errorf("failed to mark flag as hidden: %w", err)
 	}
-	server.MustRegisterCacheFlags(cmd.Flags(), "ns-cache", &config.NamespaceCacheConfig, namespaceCacheDefaults)
+	server.MustRegisterCacheFlags(namespaceCacheFlags, "ns-cache", &config.NamespaceCacheConfig, namespaceCacheDefaults)
 
-	cmd.Flags().BoolVar(&config.EnableExperimentalWatchableSchemaCache, "enable-experimental-watchable-schema-cache", false, "enables the experimental schema cache which makes use of the Watch API for automatic updates")
-	cmd.Flags().DurationVar(&config.SchemaWatchHeartbeat, "datastore-schema-watch-heartbeat", 1*time.Second, "heartbeat time on the schema watch in the datastore (if supported). 0 means to default to the datastore's minimum.")
 
 	// Flags for parsing and validating schemas.
 	cmd.Flags().BoolVar(&config.SchemaPrefixesRequired, "schema-prefixes-required", false, "require prefixes on all object definitions in schemas")
 
-	// Flags for HTTP gateway
-	util.RegisterHTTPServerFlags(cmd.Flags(), &config.HTTPGateway, "http", "gateway", ":8443", false)
-	cmd.Flags().StringVar(&config.HTTPGatewayUpstreamAddr, "http-upstream-override-addr", "", "Override the upstream to point to a different gRPC server")
-	if err := cmd.Flags().MarkHidden("http-upstream-override-addr"); err != nil {
-		return fmt.Errorf("failed to mark flag as hidden: %w", err)
-	}
-	cmd.Flags().StringVar(&config.HTTPGatewayUpstreamTLSCertPath, "http-upstream-override-tls-cert-path", "", "Override the upstream TLS certificate")
-	if err := cmd.Flags().MarkHidden("http-upstream-override-tls-cert-path"); err != nil {
-		return fmt.Errorf("failed to mark flag as hidden: %w", err)
-	}
-	cmd.Flags().BoolVar(&config.HTTPGatewayCorsEnabled, "http-cors-enabled", false, "DANGEROUS: Enable CORS on the http gateway")
-	if err := cmd.Flags().MarkHidden("http-cors-enabled"); err != nil {
-		return fmt.Errorf("failed to mark flag as hidden: %w", err)
-	}
-	cmd.Flags().StringSliceVar(&config.HTTPGatewayCorsAllowedOrigins, "http-cors-allowed-origins", []string{"*"}, "Set CORS allowed origins for http gateway, defaults to all origins")
-	if err := cmd.Flags().MarkHidden("http-cors-allowed-origins"); err != nil {
-		return fmt.Errorf("failed to mark flag as hidden: %w", err)
-	}
-
+	dispatchFlags := nfs.FlagSet("Dispatch")
 	// Flags for configuring the dispatch server
-	util.RegisterGRPCServerFlags(cmd.Flags(), &config.DispatchServer, "dispatch-cluster", "dispatch", ":50053", false)
-	server.MustRegisterCacheFlags(cmd.Flags(), "dispatch-cache", &config.DispatchCacheConfig, dispatchCacheDefaults)
-	server.MustRegisterCacheFlags(cmd.Flags(), "dispatch-cluster-cache", &config.ClusterDispatchCacheConfig, dispatchClusterCacheDefaults)
+	util.RegisterGRPCServerFlags(dispatchFlags, &config.DispatchServer, "dispatch-cluster", "dispatch", ":50053", false)
```

Comment:
this is good but I think we could further break down the dispatch flags if we wanted to: dispatch server, dispatch (client) cache, dispatch (server) cache, hashring config, etc.